src/library/itunes/itunesdao.h

```diff
@@ -0,0 +1,70 @@
+#pragma once
+
+#include <QHash>
+#include <QSqlDatabase>
+#include <QSqlQuery>
+#include <QString>
+#include <gsl/pointers>
+#include <map>
+
+#include "library/dao/dao.h"
+#include "library/itunes/itunespathmapping.h"
+#include "library/treeitem.h"
+
+const int kRootITunesPlaylistId = -1;
+
+struct ITunesTrack {
+    int id;
+    QString artist;
+    QString title;
+    QString album;
+    QString albumArtist;
+    QString genre;
+    QString grouping;
+    int year;
+    int duration;
+    QString location;
+    int rating;
+    QString comment;
+    int trackNumber;
+    int bpm;
+    int bitrate;
+};
+
+struct ITunesPlaylist {
+    int id;
+    QString name;
+};
+
+/// A wrapper around the iTunes database tables. Keeps track of the
+/// playlist tree, deals with duplicate disambiguation and can export
+/// the tree afterwards.
+class ITunesDAO : public DAO {
+  public:
+    ~ITunesDAO() override = default;
+
+    void initialize(const QSqlDatabase& database) override;
+
+    virtual bool importTrack(const ITunesTrack& track);
+    virtual bool importPlaylist(const ITunesPlaylist& playlist);
```

Comment:
I don't like the fact that these have to be virtual just in order to be tested. Is there no feature in gtest that can avoid that using some linker hacks?

Reply:
I've thought about this and the other option would be to use templates and parameterize the importer over the DAO. Unfortunately, that would also mean that we'd have to move the implementation into headers etc.

Since virtual methods are generally not thaat expensive (I've tested it on a library with hundreds of playlists and >10k tracks with pretty much no noticeable performance difference), I think this should be fine. The database calls are likely more an order of magnitude more expensive than the virtual function calls.

Reply:
Right, in the scope of everything else, those virtual calls are probably going to be cheap, I'm just used to them being expensive in the context of cache latency and branch prediction.